packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,374 @@
+import type { Reference } from '@typescript-eslint/scope-manager';
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import type { MakeRequired } from '../util';
+import { createRule, getParserServices } from '../util';
+
+type NodeWithTypeParameters = MakeRequired<
+  ts.SignatureDeclaration | ts.ClassLikeDeclaration,
+  'typeParameters'
+>;
+
+export default createRule({
+  defaultOptions: [],
+  meta: {
+    docs: {
+      description: 'Disallow type parameters that only appear once',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      sole: 'Type parameter {{name}} is used only once.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  name: 'no-unnecessary-type-parameters',
+  create(context) {
+    const parserServices = getParserServices(context);
+    return {
+      [[
+        'ArrowFunctionExpression[typeParameters]',
+        'ClassDeclaration[typeParameters]',
+        'ClassExpression[typeParameters]',
+        'FunctionDeclaration[typeParameters]',
+        'FunctionExpression[typeParameters]',
+        'TSCallSignatureDeclaration[typeParameters]',
+        'TSConstructorType[typeParameters]',
+        'TSDeclareFunction[typeParameters]',
+        'TSEmptyBodyFunctionExpression[typeParameters]',
+        'TSFunctionType[typeParameters]',
+        'TSMethodSignature[typeParameters]',
+      ].join(', ')](node: TSESTree.FunctionLike): void {
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(
+          node,
+        ) as NodeWithTypeParameters;
+
+        const checker = parserServices.program.getTypeChecker();
+        let counts: Map<ts.Identifier, number> | undefined;
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const esTypeParameter =
+            parserServices.tsNodeToESTreeNodeMap.get<TSESTree.TSTypeParameter>(
+              typeParameter,
+            );
+          const scope = context.sourceCode.getScope(esTypeParameter);
+
+          // Quick path: if the type parameter is used multiple times in the AST,
+          // we don't need to dip into types to know it's repeated.
+          if (isTypeParameterRepeatedInAST(esTypeParameter, scope.references)) {
```

Comment:
From https://github.com/typescript-eslint/typescript-eslint/pull/8173#issuecomment-2068228776, I reduced the problem a bit to:

```ts
type ValueNulls<U extends string> = {} & {
  [P in U]: null;
};

declare function invert<T extends string>(obj: T): ValueNulls<T>;
```

The issue was really interesting:

1. In the type system, `ValueNulls<T>` gets collapsed into the equivalent of `{ [P in T]: null }` (since `{} & ` goes away in intersections with `{ ... }`)
2. Since it's a type alias, the type system was considering it like `{ [P in U]: null }`, where the `U` isn't the same as the original `T`
3. ...so `T`'s total usage count was 2, instead of 3

I think there are two reasonable ways we could overcome this:

* In the type system, figure out when this kind of misdirection is happening, and adjust the reference counting logic
* Do a "quick pass" ahead of time that just uses the scope manager / AST nodes to check for having references

That "quick pass" is a lot simpler of logic and has the added benefit of avoiding potentially expensive type system calls. Win-win. And it's what you'd originally had as the design before I trimmed things down! Added back in.

Reply:
Interesting, thanks for digging into this. This definitely isn't how I'd write this type today, but that's no reason it can't be a test case!

> ...so T's total usage count was 2, instead of 3

Do you mean 1 instead of 2? Both 2 and 3 should be OK.

So the idea with the "quick pass" is that we see `ValueNulls<T>` and assume it uses `T` at least once. Since the count comes out to 2 under this assumption, we don't need to resolve any more types. That makes sense to me. If a type alias doesn't use all of its type parameters, then that should be a different error (from a different rule).

Reply:
I meant (initial) + (1 usage instead of 2), so yes 😄 that.